### PR TITLE
Updated and documented PowerShell scripts

### PR DESCRIPTION
**Pull request by CamillaCauterucci**
1. The Readme.md for the github repo should be changed to describe what is in it (i.e. general description, docs and video, script list with single phrase description)

### CODE REVIEW

File: Src/Add-AzureDtlVM.ps1 [hunk 1]
```diff
@@ -1,64 +1,142 @@
+<#
+.SYNOPSIS 
+    This script adds the specified number of Azure virtual machines in the specified DevTest Lab.
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1]: Remove repetition 'specified' in all rest of document.

File: Src/Add-AzureDtlVM.ps1 [hunk 2]
```diff
@@ -1,64 +1,142 @@
...
+
+.DESCRIPTION
+    It allows the creation of the VMs inside a specific lab. This script needs to be scheduled for the creation of the VMs of each lab.
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1]: Change 'This script needs to be ...' to 'This script can be run either from command line or Azure Automation' in all the rest of source files.

File: Src/Add-AzureDtlVM.ps1 (hunk 1 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: Sorry, I meant 'in all the rest of the source files whenever you use this phrase'

File: Src/Add-AzureDtlVM.ps1 (hunk 2 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: Please expand on the description. I.E. 'It allows to specify most parameters of interest for a lab creation (i.e. image name, network name, etc...). It should be a paragraph with one to three phrases. Please do this for all the other descriptions of functions in the source files.

File: Src/Add-AzureDtlVM.ps1 [hunk 3]
```diff
@@ -1,64 +1,142 @@
...
+
+.PARAMETER TotalLabSize
+    Mandatory. Desired total number of VMs in the lab.
+
+.PARAMETER BatchSize
+    Optional. How many VMs to create in each batch.
+    Default 30.
+
+.PARAMETER TemplatePath
+    Optional. Path to the Deployment Template File.
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1]: Or URL of the template file when running from Azure Automation.

File: Src/Add-AzureDtlVM.ps1 [hunk 4]
```diff
@@ -1,64 +1,142 @@
...
+    Default ".\dtl_multivm_customimage.json".
+
+.PARAMETER ShutdownPath
+    Optional. Path to the Shutdown file.
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1]: Same as above.

File: Src/Add-AzureDtlVM.ps1 [hunk 5]
```diff
@@ -1,64 +1,142 @@
...
+    Optional. Shutdown time for the VMs in the lab. In form of 'HH:mm' in TimeZoneID timezone.
+    Default $ExpirationTime.
+
+.EXAMPLE
+    Add-AzureDtlVM -LabName University -VMCount 50 -ImageName "UnivImage" -TotalLabSize 200
+
+.EXAMPLE
+    Add-AzureDtlVM -LabName University -VMCount 15 -ImageName "UnivImage" -TotalLabSize 15 -ExpirationTime "16:00" -DaysToExpiry 0
+
+.EXAMPLE
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1:78]: Add an example of running from Azure Automation. Look at the automation variables and pass those as parameters.

File: Src/Add-AzureDtlVM.ps1 [hunk 6]
```diff
@@ -220,14 +288,14 @@ try {
     [array] $failed = $vms | ? { $_.Properties.provisioningState -eq 'Failed' }
     LogOutput "Detected $($failed.Count) failed VMs"
 
-    RemoveBatchVms -vms $failed -batchSize $removeBatchSize -profilePath $profilePath -credentialsKind $credentialsKind
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1:223]: Why removing profilePath? It seems that it should be used in the File branch inside the RemoveBatchVms function. It should be passed into Remove-AzureDtlLabVMs as a parameter.

File: Src/Add-AzureDtlVMAutoVar.ps1 [hunk 7]
```diff
@@ -1,30 +1,72 @@
+<#
+.SYNOPSIS 
+    This script adds the specified number of Azure virtual machines in the specified DevTest Lab.
```
Comment (lucabol) [inline Src/Add-AzureDtlVMAutoVar.ps1]: 'by reading some variable from ...' to keep it to one phrase.

File: Src/Add-AzureDtlVM.ps1 [hunk 8]
```diff
@@ -1,64 +1,142 @@
...
+    Mandatory. Name of Lab.
+
+.PARAMETER VMCount
+    Mandatory. Number of VMs to create with this execution.
+
+.PARAMETER ImageName
+    Mandatory. Name of base image in lab.
+
+.PARAMETER TotalLabSize
+    Mandatory. Desired total number of VMs in the lab.
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1]: Add what this is used for, i.e. if the lab already contains N machines, it won't create more.

File: Src/Add-AzureDtlVMAutoVar.ps1 [hunk 9]
```diff
@@ -1,30 +1,72 @@
...
+    Mandatory. Name of Lab.
+
+.PARAMETER VMCount
+    Mandatory. Number of VMs to create with this execution.
+
+.PARAMETER ImageName
+    Mandatory. Name of base image in lab.
+
+.PARAMETER TotalLabSize
+    Mandatory. Desired total number of VMs in the lab.
```
Comment (lucabol) [inline Src/Add-AzureDtlVMAutoVar.ps1]: Change as above.

File: Src/DeallocateStoppedVM.ps1 [hunk 10]
```diff
@@ -1,10 +1,33 @@
-﻿Param
+﻿<#
+.SYNOPSIS 
+    This script deallocates every Azure virtual machines which are stopped.
```
Comment (lucabol) [inline Src/DeallocateStoppedVM.ps1]: deallocates every stopped Azure ...

File: Src/DeallocateStoppedVM.ps1 [hunk 11]
```diff
@@ -1,10 +1,33 @@
...
+
+.DESCRIPTION
+    This script deallocates every Azure virtual machines which are stopped in the specified DevTest Lab.
```
Comment (lucabol) [inline Src/DeallocateStoppedVM.ps1]: Expand comment. Read code and look at definition of Stopped (i.e. stopped but not deallocated).

File: Src/DeallocateStoppedVM.ps1 [hunk 12]
```diff
@@ -1,10 +1,33 @@
...
+
+.PARAMETER LabName
+    Mandatory. The name of the Dev Test Lab to clean up.
```
Comment (lucabol) [inline Src/DeallocateStoppedVM.ps1]: No clean up here.

File: Src/Manage-AzureDtlFixedPool.ps1 [hunk 13]
```diff
@@ -1,49 +1,108 @@
+<#
+.SYNOPSIS 
+    This script verifies whether the Virtual Machine pool of the Lab equals to the PoolSize specified as Azure Tag of Lab.
```
Comment (lucabol) [inline Src/Manage-AzureDtlFixedPool.ps1]: No, this script doesn't verify, but guarantees as stated in the DESCRIPTION.

File: Src/Manage-AzureDtlFixedPool.ps1 [hunk 14]
```diff
@@ -91,11 +151,26 @@ try {
 
     # Get size of the managed pool
     $Lab = GetLab -LabName $LabName
-    $poolSize = $Lab.Tags.PoolSize
-    if(! $poolSize) {
+    $poolSize = $null
+    # Hack for runbook
```
Comment (lucabol) [inline Src/Manage-AzureDtlFixedPool.ps1:155]: Wow. Good catch :-)

File: Src/Remove-AzureDtlLabVMs.ps1 [hunk 15]
```diff
@@ -1,28 +1,57 @@
+<#
+.SYNOPSIS 
+    This script deletes every Azure virtual machines in the specified DevTest Lab.
+
+.DESCRIPTION
+    This script deletes every Azure virtual machines in the specified DevTest Lab. It will
+    process all VMs in parallel divided into blocks.
+
+.PARAMETER ids
```
Comment (lucabol) [inline Src/Remove-AzureDtlLabVMs.ps1]: doc profilePath param

File: Src/Remove-AzureDtlLabVMs.ps1 [hunk 16]
```diff
@@ -1,28 +1,57 @@
+<#
+.SYNOPSIS 
+    This script deletes every Azure virtual machines in the specified DevTest Lab.
```
Comment (lucabol) [inline Src/Remove-AzureDtlLabVMs.ps1]: No, it deletes the ones with the passed Ids.

File: Src/Remove-AzureDtlLabVMs.ps1 [hunk 17]
```diff
@@ -47,17 +82,24 @@ try {
...
-        $jobs += Start-Job -Name $id -ScriptBlock $deleteVmBlock -ArgumentList $id, $profilePath
-        LogOutput "$id deleted."
+    foreach ($id in $ids) {
+        if ($credentialsKind -eq "File") { 
+            LogOutput "Starting job to delete $id ..."
+            $jobs += Start-Job -Name $id -ScriptBlock $deleteVmBlock -ArgumentList $id, $profilePath
+            LogOutput "$id deleted."
+        }
+        else {
+            DeleteSingleVM -id $id -profilePath $profilePath
```
Comment (lucabol) [inline Src/Remove-AzureDtlLabVMs.ps1:96]: Why doesn't call RemoveBatchVMs to do it in parallel?

Comment (lucabol) [reply Src/Remove-AzureDtlLabVMs.ps1:96]: Oh wait, forget about that. I get it. this is called by RemoveBatchVMs  ...

File: Src/Remove-GroupPermissionsDevTestLab.ps1 [hunk 18]
```diff
@@ -1,21 +1,50 @@
-﻿[cmdletbinding()]
+﻿<#
+.SYNOPSIS 
+    This script removes the specified role to the AD Group in the specified DevTest Lab.
```
Comment (lucabol) [inline Src/Remove-GroupPermissionsDevTestLab.ps1]: Confusing. Removes the specific AD group from the role?

File: Src/Shutdown scripts/LoadIdleScript.ps1 [hunk 19]
```diff
@@ -1,3 +1,35 @@
...
+        This script creates a task inside Windows Task Scheduler getting the Task scheduler configuration from the xml file.
+
+    .PARAMETER StorageAccountName
+        Optional. The name of the storage account
+
+    .PARAMETER containerName
+        Optional. The name of the container
+
+    .PARAMETER SASToken
+        Optional. The name of the container
```
Comment (lucabol) [inline Src/Shutdown scripts/LoadIdleScript.ps1]: Name of SAS token.

File: Src/Shutdown scripts/ShutdownOnIdleV2.ps1 [hunk 20]
```diff
@@ -1,4 +1,34 @@
-﻿ [CmdletBinding()]
+﻿ <#
+    .SYNOPSIS
+        This script checks VM Status
```
Comment (lucabol) [inline Src/Shutdown scripts/ShutdownOnIdleV2.ps1]: No, it shutdowns the machine if the user hasn't been active.

File: Src/Test-AzureDtlVMs.ps1 [hunk 21]
```diff
@@ -1,21 +1,47 @@
+<#
+.SYNOPSIS 
+    Given LabName and LabSize, this script verifies how many Azure virtual machines are inside the DevTest Lab and throws an error inside the logs when the number is greater or lower than size +/- VMDelta. 
+
+.DESCRIPTION
+    Given LabName and LabSize, this script verifies how many Azure virtual machines are inside the DevTest Lab and throws an error inside the logs when the number is greater or lower than size +/- VMDelta. 
+
+.PARAMETER LabName
+    Mandatory. The name of the Dev Test Lab to clean up.
```
Comment (lucabol) [inline Src/Test-AzureDtlVMs.ps1]: No cleaning up.

File: TODO.txt [hunk 22]
```diff
@@ -1,11 +1,11 @@
-1. Doc for class scenario (step by step installation)
```
Comment (lucabol) [inline TODO.txt:1]: Remove this file.

File: Src/SimplifiedJSPortal/Readme.md [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
++ Sign in to the [Azure portal](https://portal.azure.com)
++ Choose your Azure AD tenant by selecting your account in the top right corner of the page
++ In the left-hand navigation pane, choose More Services, click App Registrations, and click Add
++ Follow the prompts and create a new application
++ Once the app is created, grant delegated permissions to Windows Azure Active Directory (Microsoft.Azure.ActiveDirectory) and Windows Azure Service Management API in the Required permissions Settings blade 
++ Generate a valid key and copy it.
++ Register your App URL as the Reply URL's section 
++ Edit manifest and set "oauth2AllowImplicitFlow": true
+
+## Placeholder replacement
```
Comment (lucabol) [inline Src/SimplifiedJSPortal/Readme.md:18]: Please describe what goes into this fields.

**Review by lucabol — CHANGES_REQUESTED** [review]
Good job guys. The code looks good. Take a look at the comments. I'll send comments on the doc and videos separately.

File: Src/Add-AzureDtlVM.ps1 (hunk 1 above)
Comment (CamillaCauterucci) [reply Src/Add-AzureDtlVM.ps1]: Done

File: Src/Add-AzureDtlVM.ps1 (hunk 8 above)
Comment (CamillaCauterucci) [reply Src/Add-AzureDtlVM.ps1]: Done

File: Src/Add-AzureDtlVM.ps1 (hunk 3 above)
Comment (CamillaCauterucci) [reply Src/Add-AzureDtlVM.ps1]: Done

File: Src/Add-AzureDtlVM.ps1 (hunk 4 above)
Comment (CamillaCauterucci) [reply Src/Add-AzureDtlVM.ps1]: Done

File: Src/Add-AzureDtlVMAutoVar.ps1 (hunk 7 above)
Comment (CamillaCauterucci) [reply Src/Add-AzureDtlVMAutoVar.ps1]: Done

File: Src/DeallocateStoppedVM.ps1 (hunk 10 above)
Comment (CamillaCauterucci) [reply Src/DeallocateStoppedVM.ps1]: Done

File: Src/DeallocateStoppedVM.ps1 (hunk 11 above)
Comment (CamillaCauterucci) [reply Src/DeallocateStoppedVM.ps1]: Done

File: Src/DeallocateStoppedVM.ps1 (hunk 12 above)
Comment (CamillaCauterucci) [reply Src/DeallocateStoppedVM.ps1]: Modified.

File: Src/Manage-AzureDtlFixedPool.ps1 (hunk 13 above)
Comment (CamillaCauterucci) [reply Src/Manage-AzureDtlFixedPool.ps1]: Modified

File: Src/Remove-AzureDtlLabVMs.ps1 (hunk 16 above)
Comment (CamillaCauterucci) [reply Src/Remove-AzureDtlLabVMs.ps1]: Modified

File: Src/Remove-AzureDtlLabVMs.ps1 (hunk 15 above)
Comment (CamillaCauterucci) [reply Src/Remove-AzureDtlLabVMs.ps1]: Done

File: Src/Shutdown scripts/LoadIdleScript.ps1 (hunk 19 above)
Comment (CamillaCauterucci) [reply Src/Shutdown scripts/LoadIdleScript.ps1]: Modified

File: TODO.txt (hunk 22 above)
Comment (CamillaCauterucci) [reply TODO.txt:1]: Removed

File: Src/Add-AzureDtlVM.ps1 (hunk 1 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: I don't see these changed in your check-in. Just see current pharasing.

Comment (CamillaCauterucci) [reply Src/Add-AzureDtlVM.ps1]: Just did it.

File: Src/Remove-AzureDtlLabVMs.ps1 (hunk 15 above)
Comment (lucabol) [reply Src/Remove-AzureDtlLabVMs.ps1]: Verified.

File: TODO.txt (hunk 22 above)
Comment (lucabol) [reply TODO.txt:1]: Verified.

File: Src/Add-AzureDtlVM.ps1 [hunk 24]
```diff
@@ -288,7 +288,7 @@ try {
     [array] $failed = $vms | ? { $_.Properties.provisioningState -eq 'Failed' }
     LogOutput "Detected $($failed.Count) failed VMs"
 
-    RemoveBatchVms -vms $failed -batchSize $removeBatchSize -credentialsKind $credentialsKind
+    RemoveBatchVms -vms $failed -batchSize $removeBatchSize -credentialsKind $credentialsKind -profilePath $profilePath
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1]: Have you tested that it still works after passing the parameter in? Who removed in the first place and why?

File: Src/Common.ps1 [hunk 25]
```diff
@@ -343,11 +343,15 @@ function RemoveBatchVMs {
...
+    .PARAMETER profilePath
+        Optional. Path to file with Azure Profile.
+        Default "$env:APPDATA\AzProfile.txt".
+
     .NOTES
 
     #>
     [CmdletBinding()]
-    param($vms, $BatchSize, $credentialsKind)
+    param($vms, $BatchSize, $credentialsKind, $profilePath)
```
Comment (lucabol) [inline Src/Common.ps1]: But then profilePath is not used i the function (i.e. it is not passed to Remove-AzureDtlLabVMs. If we reintroduce this parameter, we have to understand why it was removed, if it is ok to put it back and test it at least once.

File: Src/Remove-AzureDtlLabVMs.ps1 (hunk 17 above)
Comment (lucabol) [reply Src/Remove-AzureDtlLabVMs.ps1:96]: Verified.

File: Src/Shutdown scripts/ShutdownOnIdleV2.ps1 (hunk 20 above)
Comment (lucabol) [reply Src/Shutdown scripts/ShutdownOnIdleV2.ps1]: Verified.

File: Src/Test-AzureDtlVMs.ps1 (hunk 21 above)
Comment (lucabol) [reply Src/Test-AzureDtlVMs.ps1]: Verified.

File: Src/Add-AzureDtlVM.ps1 [hunk 26]
```diff
@@ -1,9 +1,9 @@
 <#
 .SYNOPSIS 
-    This script adds the specified number of Azure virtual machines in the specified DevTest Lab.
+    This script adds the number of Azure virtual machines in the DevTest Lab.
```
Comment (lucabol) [inline Src/Add-AzureDtlVM.ps1]: This scripts adds the specified number of Azure virtual machines to a DevTest Lab.

File: Src/Add-GroupPermissionsDevTestLab.ps1 [hunk 27]
```diff
@@ -1,6 +1,6 @@
 ﻿<#
 .SYNOPSIS 
-    This script adds the specified role to the AD Group in the specified DevTest Lab.
+    This script adds the role to the AD Group in the DevTest Lab.
```
Comment (lucabol) [inline Src/Add-GroupPermissionsDevTestLab.ps1]: This script adds the specified role to the AD Group in the DevTest Lab.

File: Src/Add-AzureDtlVM.ps1 (hunk 2 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: Verified.

File: Src/Add-AzureDtlVM.ps1 (hunk 8 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: Verified.

File: Src/Add-AzureDtlVM.ps1 (hunk 3 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: Verified.

File: Src/Add-AzureDtlVM.ps1 (hunk 4 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: Verified.

File: Src/Add-AzureDtlVMAutoVar.ps1 (hunk 7 above)
Comment (lucabol) [reply Src/Add-AzureDtlVMAutoVar.ps1]: Verified.

File: Src/Add-AzureDtlVMAutoVar.ps1 (hunk 9 above)
Comment (lucabol) [reply Src/Add-AzureDtlVMAutoVar.ps1]: Verified.

File: Src/DeallocateStoppedVM.ps1 (hunk 10 above)
Comment (lucabol) [reply Src/DeallocateStoppedVM.ps1]: Verified.

File: Src/DeallocateStoppedVM.ps1 (hunk 11 above)
Comment (lucabol) [reply Src/DeallocateStoppedVM.ps1]: Verified.

File: Src/DeallocateStoppedVM.ps1 (hunk 12 above)
Comment (lucabol) [reply Src/DeallocateStoppedVM.ps1]: Verified.

File: Src/Manage-AzureDtlFixedPool.ps1 (hunk 13 above)
Comment (lucabol) [reply Src/Manage-AzureDtlFixedPool.ps1]: Verified.

File: Src/Manage-AzureDtlFixedPool.ps1 (hunk 14 above)
Comment (lucabol) [reply Src/Manage-AzureDtlFixedPool.ps1:155]: Verified.

File: Src/Remove-AzureDtlLabVMs.ps1 (hunk 16 above)
Comment (lucabol) [reply Src/Remove-AzureDtlLabVMs.ps1]: Verified.

File: Src/Remove-GroupPermissionsDevTestLab.ps1 (hunk 18 above)
Comment (lucabol) [reply Src/Remove-GroupPermissionsDevTestLab.ps1]: Verified.

File: Src/Shutdown scripts/LoadIdleScript.ps1 (hunk 19 above)
Comment (lucabol) [reply Src/Shutdown scripts/LoadIdleScript.ps1]: Verified.

File: Src/SimplifiedJSPortal/Readme.md (hunk 23 above)
Comment (CamillaCauterucci) [reply Src/SimplifiedJSPortal/Readme.md:18]: Done

Comment (lucabol) [reply Src/SimplifiedJSPortal/Readme.md:18]: Verified.

File: Src/Add-AzureDtlVM.ps1 (hunk 5 above)
Comment (CamillaCauterucci) [reply Src/Add-AzureDtlVM.ps1:78]: Done

Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1:78]: This wouldn't work in Azure Automation as it reference files. It should reference URLs. That's the main difference.

File: Src/Readme.md [hunk 28]
```diff
@@ -0,0 +1,45 @@
+# Source folder
```
Comment (lucabol) [inline Src/Readme.md]: We need just one Readme.md at the University level (one level up), that describes everything in the repo. Usually there are no Readme.md in the Src folders. Also, please use MarkDown for formatting. If you look at this link the functions don't come out nicely formatted: https://github.com/lucabol/University/tree/FixesToCheck/Src

File: Readme.md [hunk 29]
```diff
@@ -1,25 +1,15 @@
...
-## Creating the appropriate Azure credential file to run the script in Class
-In 'powershell' do the following:
-
-    Login-AzureRmAccount
-    Set-AzureRmContext -SubscriptionId "XXXXX-XXXX-XXXX"
-    Save-AzureRMProfile -Path "$env:APPDATA\AzProfile.txt"
-
-This saves the credentials file where the scripts look for.
\ No newline at end of file
+# University repository
```
Comment (lucabol) [inline Readme.md:1]: Two changes for this Readme please:
1. Under each category of scripts, put the list of scripts and what each one does
2. For each item (video, doc or script) put a link that brings you directly to it

File: Src/Add-AzureDtlVM.ps1 (hunk 1 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1]: Verified.

File: Src/Add-AzureDtlVM.ps1 (hunk 5 above)
Comment (lucabol) [reply Src/Add-AzureDtlVM.ps1:78]: Verified

File: Readme.md (hunk 29 above)
Comment (CamillaCauterucci) [reply Readme.md:1]: Done

**Review by lucabol — APPROVED** [review]

**Pull request merged**